File: frontend/types/growth-center.ts
// 成长档案中心类型定义（对齐后端 app/schemas/growth.py）

/** 成长轨迹事件类型枚举 */
export type TrajectoryEventType =
  | "action_checkin"
  | "review_completed"
  | "milestone";

/** 档案状态枚举 */
export type ArchiveStatus = "ACTIVE" | "STALE";

/** 记录成长轨迹事件请求（user_id 由登录态 token 推断） */
export interface GrowthTrajectoryCreateRequest {
  event_type: TrajectoryEventType;
  /** 事件负载（打卡 / 复盘 / 里程碑明细） */
  event_payload: Record<string, unknown>;
  occurred_at: string;
  /** 上游事件幂等 ID（重复提交被丢弃） */
  source_event_id?: string | null;
}

/** 成长轨迹事件 VO */
export interface GrowthTrajectoryVO {
  id: number;
  user_id: string;
  event_type: TrajectoryEventType;
  event_payload: Record<string, unknown>;
  occurred_at: string;
  source_event_id: string | null;
}

/** 成长轨迹时间轴 VO */
export interface GrowthTrajectoryListResponse {
  items: GrowthTrajectoryVO[];
  total: number;
}

/** 档案聚合 VO */
export interface GrowthArchiveVO {
  user_id: string;
  /** 行动完成率；0.0~1.0，保留 2 位小数 */
  action_completion_rate: number;
  total_actions: number;
  completed_actions: number;
  /** 当前 Streak Days */
  streak_days: number;
  /** 加权行动完成分（D18 北极星指标） */
  weighted_action_score: number;
  archive_status: ArchiveStatus;
  /** 最近聚合时间 */
  updated_at: string;
}

/** 成长统计 VO（行动完成率 + Streak 统计） */
export interface GrowthStatsVO {
  user_id: string;
  /** 行动完成率；0.0~1.0 */
  action_completion_rate: number;
  current_streak_days: number;
  longest_streak_days: number;
  total_actions: number;
  completed_actions: number;
}

File: backend/app/models/growth_center.py
"""成长档案中心契约模型（系统设计 §4.2.5 ~ §4.2.6）。

MVP 方案 C：契约先行、实现延后 — 仅落库建表，业务逻辑后续实现。
"""

from datetime import datetime
from uuid import UUID

from sqlalchemy import DateTime, Index, Integer, Numeric, String, UniqueConstraint, text
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base
from app.models.base import GUID, JSONB, BigIntPK, ContractAuditMixin


class GrowthTrajectory(ContractAuditMixin, Base):
    """成长轨迹事件流（t_growth_trajectory）。

    event_type：action_checkin / review_completed / milestone。
    event_payload：事件结构化载荷（JSONB）。
    """

    __tablename__ = "t_growth_trajectory"

    id: Mapped[int] = mapped_column(BigIntPK, primary_key=True, autoincrement=True)
    user_id: Mapped[UUID] = mapped_column(GUID(), nullable=False)
    event_type: Mapped[str] = mapped_column(String(20), nullable=False)
    # action_checkin / review_completed / milestone
    event_payload: Mapped[dict] = mapped_column(JSONB, nullable=False)
    source_event_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    occurred_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    __table_args__ = (
        UniqueConstraint("source_event_id", name="uk_growth_trajectory_source_event_id"),
        Index("idx_growth_trajectory_user_id_occurred_at", "user_id", "occurred_at"),
    )


class GrowthArchive(ContractAuditMixin, Base):
    """成长档案聚合快照（t_growth_archive）。

    archive_status：ACTIVE / STALE。
    """

    __tablename__ = "t_growth_archive"

    id: Mapped[int] = mapped_column(BigIntPK, primary_key=True, autoincrement=True)
    user_id: Mapped[UUID] = mapped_column(GUID(), nullable=False)
    action_completion_rate: Mapped[float] = mapped_column(
        Numeric(5, 2), nullable=False, default=0, server_default=text("0")
    )
    total_actions: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default=text("0")
    )
    completed_actions: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default=text("0")
    )
    streak_days: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default=text("0")
    )
    weighted_action_score: Mapped[float] = mapped_column(
        Numeric(5, 2), nullable=False, default=0, server_default=text("0")
    )
    archive_status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="ACTIVE", server_default=text("'ACTIVE'")
    )
    # ACTIVE / STALE

    __table_args__ = (UniqueConstraint("user_id", name="uk_growth_archive_user_id"),)
